File: Microsoft/102.binary-tree-level-order-traversal.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        /*
        Approach:
        Breadth-first search groups nodes by depth. For each loop, capture the
        current queue size and pop exactly that many nodes to form one level.

        C++ notes:
        queue<TreeNode*> provides FIFO traversal, matching level-order behavior.

        Complexity: O(n) time and O(w) space where w is maximum tree width.
        */
        if (!root) return {};
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int size = (int)q.size();
            vector<int> level;
            for (int i = 0; i < size; ++i) {
                TreeNode* node = q.front();
                q.pop();
                level.push_back(node->val);
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            ans.push_back(move(level));
        }
        return ans;
    }
};

File: companies/ZScaler/102.binary-tree-level-order-traversal.go
package leetcode

// LevelOrder102 is breadth-first search. A slice with a moving head index acts
// as a FIFO queue; each outer loop consumes the current queue size to form one
// depth level.
//
// Time: O(n). Space: O(w), maximum tree width.
func LevelOrder102(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	ans := [][]int{}
	q := []*TreeNode{root}
	for len(q) > 0 {
		size := len(q)
		level := make([]int, 0, size)
		for i := 0; i < size; i++ {
			node := q[0]
			q = q[1:]
			level = append(level, node.Val)
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		ans = append(ans, level)
	}
	return ans
}
